packages-content-model/roosterjs-content-model-core/lib/publicApi/table/setTableCellBackgroundColor.ts

```diff
@@ -43,28 +45,80 @@ export function setTableCellBackgroundColor(
             delete cell.format.textColor;
         }
 
-        if (applyToSegments && cell.format.textColor) {
-            cell.blocks.forEach(block => {
-                if (block.blockType == 'Paragraph') {
+        if (applyToSegments) {
+            setAdaptiveCellColor(cell);
+        }
+    } else {
+        delete cell.format.backgroundColor;
+        delete cell.format.textColor;
+        if (applyToSegments) {
+            removeAdaptiveCellColor(cell);
+        }
+    }
+
+    delete cell.cachedElement;
+}
+
+function removeAdaptiveCellColor(cell: ContentModelTableCell) {
+    cell.blocks.forEach(block => {
+        if (block.blockType == 'Paragraph') {
+            if (
+                block.segmentFormat?.textColor &&
+                shouldRemoveColor(block.segmentFormat?.textColor, cell.format.backgroundColor || '')
+            ) {
+                delete block.segmentFormat.textColor;
+            }
+            block.segments.forEach(segment => {
+                if (
+                    segment.format.textColor &&
+                    shouldRemoveColor(segment.format.textColor, cell.format.backgroundColor || '')
+                ) {
+                    delete segment.format.textColor;
+                }
+            });
+        }
+    });
+}
+
+function setAdaptiveCellColor(cell: ContentModelTableCell) {
+    if (cell.format.textColor) {
+        cell.blocks.forEach(block => {
+            if (block.blockType == 'Paragraph') {
+                if (!block.segmentFormat?.textColor) {
                     block.segmentFormat = {
                         ...block.segmentFormat,
                         textColor: cell.format.textColor,
                     };
-                    block.segments.forEach(segment => {
+                }
+                block.segments.forEach(segment => {
+                    if (!segment.format?.textColor) {
                         segment.format = {
                             ...segment.format,
                             textColor: cell.format.textColor,
                         };
-                    });
-                }
-            });
-        }
-    } else {
-        delete cell.format.backgroundColor;
-        delete cell.format.textColor;
+                    }
+                });
+            }
+        });
     }
+}
 
-    delete cell.cachedElement;
+/**
+ * If the cell background color is white or black, and the text color is white or black, we should remove the text color
+ * @param textColor the segment or block text color
+ * @param cellBackgroundColor the cell background color
+ * @returns
+ */
+function shouldRemoveColor(textColor: string, cellBackgroundColor: string) {
+    if (
+        ([White, 'rgb(255,255,255)'].indexOf(textColor) > -1 &&
```

Comment:
It seems you are checking both text color and background color to be white (or black). Are you sure about that?
It only returns true when black on black, or white on white.